### PR TITLE
Use Identifiers instead of Strings 

### PR DESCRIPTION
**Pull request by ttsugriy**
The high level summary of the changes:
- canonicalize identifier names so that their identity could be used for equality checks. This also should improve memory usage.
- use `Identifier` instead of `name` in `Keyword` and `Parameter`.

These changes are in preparation for using `Identifier` instead of `name` for environment lookups. Technically it could be possible to achieve the same by interning all names, but it would be hard to maintain this invariant in an efficient way.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
 package com.google.devtools.build.lib.syntax;
 
 import com.google.devtools.build.lib.util.SpellChecker;
+import com.google.devtools.build.lib.util.StringCanonicalizer;
```
Comment (alandonovan) [inline src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: I haven't looked ahead in this stack of PRs yet, but:

Neither identifiers nor strings should be used as the key to an environment lookup, at least in an efficient (flat) representation. It is the job of the validation pass to relate each Identifier to the entity (in Skylark, always a variable) to which it refers, which is either a global, a local, a free variable, or a builtin. For each of those categories, and especially for the last two kinds, the validation pass also computes a dense numbering, so that variables can be identified by an index and accessed by array indexing. I suggest you follow the Go implementation:

https://github.com/google/skylark/blob/master/resolve/resolve.go
https://github.com/google/skylark/blob/master/syntax/syntax.go#L256

String interning is not relevant: equal but nonidentical strings (e.g. from different files) may denote the same variable; conversely, identical strings may denote different variables. Nor, I suspect, is it even beneficial, but of course that question is easily settled by profiling data.

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java [hunk 2]
```diff
@@ -61,7 +62,9 @@ public void prettyPrint(Appendable buffer) throws IOException {
   public boolean equals(@Nullable Object object) {
     if (object instanceof Identifier) {
       Identifier that = (Identifier) object;
-      return this.name.equals(that.name);
+      // use identity check for improved performance as long as all names are canonicalized
+      // This is currently ensured by the static factory method creating Identifiers.
+      return this.name == that.name;
```
Comment (alandonovan) [inline src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: Interning is entirely unnecessary for the correctness of the name resolution algorithm. Does this function show up in a profile? I would be astonished if so; why are identifiers ever compared for equality?

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 1 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: @alandonovan, you're correct that ideally an array should be used instead of dictionary, but I'm following https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/syntax/Identifier.java;l=23?q=Identifier.java which seems to be a good temporary compromise until we get to most efficient implementation based on arrays. Thanks for the pointers, btw!

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 2 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: because, since identifier names are now interned, their equality can now be checked using identity check instead of a value equality check

**Comment by ttsugriy** [comment]
thanks for the pointers, @alandonovan!

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 1 above)
Comment (alandonovan) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: The linked comment says "for extra performance". Does this change actually result in extra performance? If not, what is the benefit of requiring that Identifier names be canonical?

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 2 above)
Comment (alandonovan) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: But how does that help? If the answer is "performance", then that should be supported by evidence because (a) the change adds a cost, namely, the type of Identifier.name is no longer String but the more complex type "String interned by a particular interner", and (b) it is far from obvious that the overhead of interning these strings will be successfully amortized by subsequent cheaper equality tests.

Why does the performance of Identifier.equals matter anyway?

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 1 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: @alandonovan, as I've mentioned above, variable lookups have been called out as a bottleneck in a comment - https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/syntax/Identifier.java;l=27?q=Identifier.java

The improved performance in this case means specifically the performance of `==` vs `equals` which is obviously better. The question of whether it improves all use cases is a more complex one and so far I only tested this with only a few different scenarios where there does not seem to be a pronounced difference.

I would like to emphasize that this interning is not the main purpose of this PR, but just a tiny attempt for optimization. Since it clearly leads to some controversy, I'll remove this change and submit a separate PR with just this change to not derail the conversation from the main point of using `Identifier`s instead of `String`s.

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 2 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: @alandonovan, the reason why I've added this optimization was solely based on a comment from the original authors https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/syntax/Identifier.java;l=27?q=Identifier.java and is not the main purpose of this change, so as I've said above, I can separate this change so that it can be discussed separately.

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 1 above)
Comment (alandonovan) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: > The improved performance in this case means specifically the performance of == vs equals which is obviously better. 

The performance of == may be obviously better, but the performance of == combined with StringCanonicalizer is not obviously better or worse; nothing about it is obvious.

My whole point here may be summarized as: changes that purport to be optimizations must be justified by empirical data, not speculation. Program dynamics cannot be deduced from the source code alone.

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 2 above)
Comment (alandonovan) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: That seems best. The main point of the change, using Identifiers instead of Strings, is a good step.

**Comment by ttsugriy** [comment]
removed controversial change :)

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java [hunk 3]
```diff
@@ -117,4 +117,9 @@ EvalException createInvalidIdentifierException(Set<String> symbols) {
     String suggestion = SpellChecker.didYouMean(name, symbols);
     return new EvalException(getLocation(), "name '" + name + "' is not defined" + suggestion);
   }
+
+  /** @return The {@link Identifier} of the provided name. */
+  public static Identifier of(String name) {
```
Comment (alandonovan) [inline src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: Now that it no longer canonicalizes the string, this method is no longer necessary, nor all the updates to the call sites.

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 1 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: I totally agree with you, @alandonovan. Since my personal benchmarks showed only a marginal improvements (2-3%), pretty much the only reason why I made this change was just because it was very easy and I relied on the judgement of the comment author, but you're right that it's not the right way to approach optimizations :)

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 3 above)
Comment (ttsugriy) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: even though it's not technically necessary I believe that it makes the code look more fluent (`Identifier.of('some_variable')` which reads like English vs `new Identifier('some_variable')`) and it's also a common practice for important library code, which it makes code evolution much easier as it makes it easier to maintain open/closed design principle - for instance if one of the use cases can be handled more efficiently by having a specialized `Identifier` (like the one for `Star` or `StarStart` for example), it's possible to return that specialized instance from the factory method, but if constructor is used, all the ifs would have to go inside of the class itself. Also, in case things like interning do end up being useful, it would be easy to do this in a factory method, whereas constructor always returns a new instance of the object.

In other words, I see only advantages of using factory method and no downsides (JIT will inline it anyways, so no perf overhead). Do you happen to have some concerns about usage of a factory method?

Comment (alandonovan) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: Reading like English is not a goal. A programmer should be as fluent in their main programming language as in their mother tongue. The meaning of `new T(x)` needs no explanation, whereas `T.of(x)` is another function one must read and remember, even if it is just one drop of needless abstraction in an existing ocean. I read many wondrous claims about the JVM making the costs of abstraction disappear, but few supported by measurement. Blaze spends 2 seconds of CPU before control reaches main, largely due to code size.

I'll defer to the Bazel team for style issues though.

Comment (laurentlb) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: I think this makes little difference. I'm fine with either keeping or reverting it.

File: src/main/java/com/google/devtools/build/lib/syntax/Identifier.java (hunk 1 above)
Comment (alandonovan) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: > personal benchmarks showed only a marginal improvements (2-3%)

Now we're talking about empirical data---good. Could you elaborate on what you measured and what you found? A 2-3% improvement is no small thing.

I still don't understand why Identifier even needs an equals method.  Is it necessary and does its performance matter? Or is the observed performance gain a result of reducing persistent memory allocation?  (You could measure the effects independently by keeping StringCanonicalizer but reverting the Identifier.equals change.)

Comment (laurentlb) [reply src/main/java/com/google/devtools/build/lib/syntax/Identifier.java]: Identifier is used as a key in a map (https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/syntax/LoadStatement.java) and a set (https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/syntax/LValue.java).

In both cases, it looks like we could use a list instead (we don't need fast lookup).

**Comment by alandonovan** [comment]
On 5 June 2018 at 11:12, Laurent Le Brun <notifications@github.com> wrote:

> Identifier is used as a key in a map (https://github.com/
> bazelbuild/bazel/blob/master/src/main/java/com/google/
> devtools/build/lib/syntax/LoadStatement.java) and a set (
> https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/
> devtools/build/lib/syntax/LValue.java).
>
> In both cases, it looks like we could use a list instead (we don't need
> fast lookup).
>
Indeed. Maps are a strange data type to use in a syntax tree, because they
raise questions of ordering and uniqueness. I recommend you use lists, and
do away with Identifier.equals.  Later passes can build maps if they really
need it.

**Comment by ttsugriy** [comment]
@laurentlb, so are there any concerns with merging this PR? I do plan to work on further optimizations once this change is merged.

**Comment by laurentlb** [comment]
Importing
